### DISCUSSION TITLE
transform stylex.create() and stylex.props clientside

### DISCUSSION THREAD

**Discussion by philholden** [facebook/stylex#711] [General]
I am currently using Emotion and considering moving to StyleX the biggest blocker is getting StyleX working in our user editable playgrounds. Just need the ability to ability create and apply classes but do not need the ability to define variables. The theme variables can be available precompiled on on the playground scope but I want users of the playground to be able to override simple styles like margin on components.

The editor has a transformCode prop allowing the code entered to be transformed so ideally I'd like to use that to transform the code with a subset of stylex. Looking to be able to do simple things like this:

<img width="627" alt="image" src="https://github.com/user-attachments/assets/208974b3-c507-4ca3-ba73-65af8e91bc59">

However the babel transform code has lots of node ecosystem stuff in it which stops it running in the browser. The playground does not allow importing or exporting so does needs very reduced stylex functionality.

**Comment by nmn** [comment]
You should be able to use `@stylexjs/dev-runtime`. It's a fully runtime version of StyleX that works without compilation. It computes and injects styles at runtime.

Generally this is only suggested for development, but it might work for your use-case as well.

- [Getting started with dev-runtime](https://stylexjs.com/docs/learn/installation/#local-development-only)
- [API Docs](https://stylexjs.com/docs/api/configuration/dev-runtime/)

**Reply by philholden** [reply]
Thanks when I call inject in my index.ts

```js
import inject from '@stylexjs/dev-runtime';

inject({
  classNamePrefix: 'x',
  dev: true,
  test: false,
});
```

I get Uncaught TypeError (0, import_dev_runtime.default) is not a function. This happens on calling inject. Calling `inject.default({
  classNamePrefix: 'x',
  dev: true,
  test: false,
})` did not error. But I still got an error saying stylex.create should never be called at runtime. It should be compiled away by '@stylexjs/babel-plugin'

Here is an example in codesandbox:

https://codesandbox.io/p/sandbox/567392

```js
// index.js
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import inject from "@stylexjs/dev-runtime";
inject({
  classNamePrefix: "x",
  dev: true,
  test: false,
});

import App from "./App";

const rootElement = document.getElementById("root");
const root = createRoot(rootElement);

root.render(
  <StrictMode>
    <App />
  </StrictMode>
);
```

```js
// App.js
import "./styles.css";
import * as stylex from "@stylexjs/stylex";

const styles = stylex.create({
  foo: { color: "red" },
});

export default function App() {
  return (
    <div className="App">
      <h1 {...stylex.props(styles.foo)}>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
    </div>
  );
}
```

<img width="535" alt="image" src="https://github.com/user-attachments/assets/782ce9a5-eb6a-4d53-92ac-298f309ffa00">

**Reply by nmn** [reply]
Codesandbox was super buggy for me but I tried StackBlitz and was able to reproduce the issue. At least with Vite, the `dev-runtime` ends up pulling in a separate second copy of StyleX and patches that *unrelated* copy of StyleX which creates this problem.

I'll update the `dev-runtime` package so it's possible to import and use all of stylex directly from that package instead of just monkey patching `@stylexjs/stylex` which is what we do today.

**Reply by philholden** [reply]
Thanks so much that would be so helpful.

**Reply by nmn** [reply]
I made this change in the last release almost a month ago. Sorry I didn't update you here. Please let me know if the new implementation works for you.
